docs/getting-started.rst

```diff
@@ -0,0 +1,249 @@
+recommend some conventions here. Namely:
+
+* ``resources`` should be a mapping where the keys are the Pulumi type-strings for the resources they are configuring.
+  For example, if you want to build a VPC with several subnets, you might use the ``tb_pulumi.network.MultiCidrVpc``
+  class. Following this convention, that should be accompanied by a ``tb:network:MultiCidrVpc`` key in this mapping.
+* The values these keys map to should themselves be mappings. This provides a convention where more than one of each
+  pattern are configurable. The keys here should be arbitrary but unique identifiers for the resources being configured.
+  F/ex: ``backend`` or ``api``.
+* The values these keys map to should be a mapping where the keys are valid configuration options for the resources
+  being built. The full listing of these values can be found by browsing the documentation.
+
+
+Define a ThunderbirdPulumiProject
+"""""""""""""""""""""""""""""""""
+
+In your ``__main__.py`` file, start with a simple skeleton (or use ``__main__.py.example`` to start):
+::
+
+  import tb_pulumi
+  project = tb_pulumi.ThunderbirdPulumiProject()
+
+If you have followed the conventions outlined above, ``project`` is now an object with a key property, ``config``, which
+gives you access to the config file's data. You can use this in the next step to feed parameters into resource
+declarations.
+
+
+Declare ThunderbirdComponentResources
+"""""""""""""""""""""""""""""""""""""
+
+A ``pulumi.ComponentResource`` is a collection of related resources. In an effort to follow consistent patterns across
+infrastructure projects, the resources available in this module all extend a custom class called a
+``ThunderbirdComponentResource``. If you have followed the conventions outlined so far, it should be easy to stamp out
+common patterns with them by passing config options into the constructors for these classes.
+
+.. note::
+   The `Quickstart`_ section provides a working minimal example of code that follows these patterns.
+
+Implementing ThunderbirdComponentResources
+""""""""""""""""""""""""""""""""""""""""""
+
+So you want to develop a new pattern to stamp out? Here's what you'll need to do:
+
+* Determine the best place to put the code. Is there an existing module that fits the bill?
+* Determine the Pulumi type string for it. This goes: ``org:module:class``. The ``org`` should be unique to your
+  organization. For Thunderbird projects, it should be ``tb``. The ``module`` will be the Python submodule you're
+  placing the new class in. The ``class`` is whatever you've called the class.
+* Design the class following these guidelines:
+    * The constructor should always accept, before any other arguments, the following positional options:
+        * ``name``: The internal name of the resource as Pulumi tracks it.
+        * ``project``: The ThunderbirdPulumiProject these resources belong to.
+    * The constructor should always accept the following keyword arguments:
```

Comment:
is opts not positional? I'm not sure why this needs to be stated twice. The top can be adjusted to "The constructor should always accept, in order, the following parameters:"

Reply:
No, `opts` refers to a Pulumi-specific keyword argument to Resource constructors (it refers to a [`pulumi.ResourceOptions`](https://www.pulumi.com/docs/reference/pkg/python/pulumi/#pulumi.ResourceOptions object). It has to be something we can optionally pass in to the constructor of any class extended from a [Resource](https://www.pulumi.com/docs/reference/pkg/python/pulumi/#pulumi.Resource) because at the level where we define those we want to merge "higher order" resource options common to all resources throughout a project (like resource modification protection) with lower-level resource options (like parentage, which ComponentResource a Resource belongs to). There's a merge function we [call here](https://github.com/thunderbird/pulumi/blob/main/tb_pulumi/__init__.py#L100-L103). We could probably get rid of this by fixing the code linked there to have a default empty `opts` to merge with. Then `opts` could become part of the rest of the `kwargs`, taken into account only when it's present.